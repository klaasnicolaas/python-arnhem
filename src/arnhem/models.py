"""Models for Open Data Platform of Arnhem."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass
class ParkingSpot:
    """Object representing a parking spot."""

    spot_id: int
    parking_type: str
    street: str
    traffic_sign: str

    neighborhood: str
    neighborhood_code: str
    district: str
    district_code: str
    area: str

    coordinates: list[float]

    @classmethod
    def from_json(cls: type[ParkingSpot], data: dict[str, Any]) -> ParkingSpot:
        """Return a ParkingSpot object from a JSON dictionary.

        Args:
        ----
            data: The JSON data from the API.

        Returns:
        -------
            A ParkingSpot object.
        """
        attr = data["attributes"]
        geo = data["geometry"].get("rings")[0]
        return cls(
            spot_id=attr["OBJECTID"],
            parking_type=attr["SOORT"],
            street=attr["STRAAT"],
            traffic_sign=attr["RVV_SOORT"],
            neighborhood=attr["BUURTNAAM"],
            neighborhood_code=attr["BUURTCODE"],
            district=attr["WIJKNAAM"],
            district_code=attr["WIJKCODE"],
            area=attr["GEBIED"],
            coordinates=geo,
        )
