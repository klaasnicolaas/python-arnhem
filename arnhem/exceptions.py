"""Asynchronous Python client providing Open Data information of Arnhem."""


class ODPArnhemError(Exception):
    """Generic Open Data Platform Arnhem exception."""


class ODPArnhemConnectionError(ODPArnhemError):
    """Open Data Platform Arnhem - connection exception."""


class ODPArnhemNoResultsError(ODPArnhemError):
    """Open Data Platform Arnhem - no results found exception."""
