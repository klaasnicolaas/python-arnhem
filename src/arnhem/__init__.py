"""Asynchronous Python client providing Open Data information of Arnhem."""

from .arnhem import ODPArnhem
from .exceptions import (
    ODPArnhemConnectionError,
    ODPArnhemError,
    ODPArnhemNoResultsError,
)
from .models import ParkingSpot

__all__ = [
    "ODPArnhem",
    "ODPArnhemNoResultsError",
    "ODPArnhemConnectionError",
    "ODPArnhemError",
    "ParkingSpot",
]
