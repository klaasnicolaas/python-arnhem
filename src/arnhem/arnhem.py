"""Asynchronous Python client providing Open Data information of Arnhem."""
from __future__ import annotations

import asyncio
import socket
from dataclasses import dataclass
from importlib import metadata
from typing import Any, cast

import async_timeout
from aiohttp import ClientError, ClientSession
from aiohttp.hdrs import METH_GET
from yarl import URL

from .exceptions import (
    ODPArnhemConnectionError,
    ODPArnhemError,
    ODPArnhemNoResultsError,
)
from .models import ParkingSpot


@dataclass
class ODPArnhem:
    """Main class for handling data fetchting from Open Data Platform of Arnhem."""

    request_timeout: float = 10.0
    session: ClientSession | None = None

    _close_session: bool = False

    async def _request(
        self,
        uri: str,
        *,
        method: str = METH_GET,
        params: dict[str, Any] | None = None,
    ) -> Any:
        """Handle a request to the Open Data Platform API of Arnhem.

        Args:
        ----
            uri: Request URI, without '/', for example, 'status'
            method: HTTP method to use, for example, 'GET'
            params: Extra options to improve or limit the response.

        Returns:
        -------
            A Python dictionary with the response from
            the Open Data Platform API of Arnhem.

        Raises:
        ------
            ODPArnhemConnectionError: An error occurred while
                communicating with the Open Data Platform API of Arnhem.
            ODPArnhemError: Received an unexpected response from
                the Open Data Platform API of Arnhem.
        """
        version = metadata.version(__package__)
        url = URL.build(
            scheme="https",
            host="geo.arnhem.nl",
            path="/arcgis/rest/services/",
        ).join(
            URL(uri),
        )

        headers = {
            "Accept": "application/json",
            "User-Agent": f"PythonODPArnhem/{version}",
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        try:
            async with async_timeout.timeout(self.request_timeout):
                response = await self.session.request(
                    method,
                    url,
                    params=params,
                    headers=headers,
                    ssl=True,
                )
                response.raise_for_status()
        except asyncio.TimeoutError as exception:
            msg = "Timeout occurred while connecting to the Open Data Platform API."
            raise ODPArnhemConnectionError(msg) from exception
        except (ClientError, socket.gaierror) as exception:
            msg = "Error occurred while communicating with the Open Data Platform API."
            raise ODPArnhemConnectionError(msg) from exception

        content_type = response.headers.get("Content-Type", "")
        if "application/json" not in content_type:
            text = await response.text()
            msg = "Unexpected content type response from the Open Data Platform API."
            raise ODPArnhemError(msg, {"Content-Type": content_type, "response": text})

        return cast(dict[str, Any], await response.json())

    async def locations(
        self,
        limit: int = 10,
        set_filter: str = "1=1",
    ) -> list[ParkingSpot]:
        """Get all the parking locations.

        Args:
        ----
            limit: The number of results to return.
            filter: A filter to apply to the results.

        Returns:
        -------
            A list of ParkingSpot objects.
        """
        msg: str = "No results found, check your filter."
        results: list[ParkingSpot] = []
        locations = await self._request(
            "OpenData/Parkeervakken/MapServer/0/query",
            params={
                "where": set_filter,
                "outFields": "*",
                "outSR": "4326",
                "f": "json",
                "resultRecordCount": limit,
            },
        )

        try:
            for item in locations["features"]:
                results.append(ParkingSpot.from_json(item))
        except KeyError as exception:
            # when wrong filter is given
            raise ODPArnhemNoResultsError(msg) from exception

        # when filter is incorrect/empty
        if locations["features"] == []:
            raise ODPArnhemNoResultsError(msg)
        return results

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> ODPArnhem:
        """Async enter.

        Returns
        -------
            The Open Data Platform Arnhem object.
        """
        return self

    async def __aexit__(self, *_exc_info: Any) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.
        """
        await self.close()
