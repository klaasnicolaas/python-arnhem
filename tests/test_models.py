"""Test the models."""

from __future__ import annotations

import pytest
from aresponses import ResponsesMockServer
from syrupy.assertion import SnapshotAssertion

from arnhem import (
    ODPArnhem,
    ODPArnhemNoResultsError,
    ParkingSpot,
)

from . import load_fixtures


async def test_all_locations(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    odp_arnhem_client: ODPArnhem,
) -> None:
    """Test the parking locations model."""
    aresponses.add(
        "geo.arnhem.nl",
        "/arcgis/rest/services/OpenData/Parkeervakken/MapServer/0/query",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("e6a_parking.json"),
        ),
    )
    locations: list[ParkingSpot] = await odp_arnhem_client.locations()
    assert locations == snapshot


async def test_wrong_filter(
    aresponses: ResponsesMockServer,
    odp_arnhem_client: ODPArnhem,
) -> None:
    """Test when a wrong filter is provided."""
    aresponses.add(
        "geo.arnhem.nl",
        "/arcgis/rest/services/OpenData/Parkeervakken/MapServer/0/query",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("wrong_filter.json"),
        ),
    )
    with pytest.raises(ODPArnhemNoResultsError):
        await odp_arnhem_client.locations(set_filter="wrong=filter")


async def test_incorrect_filter(
    aresponses: ResponsesMockServer,
    odp_arnhem_client: ODPArnhem,
) -> None:
    """Test when an incorrect filter is provided."""
    aresponses.add(
        "geo.arnhem.nl",
        "/arcgis/rest/services/OpenData/Parkeervakken/MapServer/0/query",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("incorrect_filter.json"),
        ),
    )
    with pytest.raises(ODPArnhemNoResultsError):
        await odp_arnhem_client.locations(set_filter="incorrect")
