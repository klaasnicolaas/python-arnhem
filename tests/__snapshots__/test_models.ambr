# serializer version: 1
# name: test_all_locations
  list([
    ParkingSpot(spot_id=96, parking_type='bijzonder', street='G.A. van Nispenstraat', traffic_sign='E6a', neighborhood='Burgemeesterswijk', neighborhood_code='BU02021261', district=None, district_code='WK020212', area='Sector C Burgermeesterkwartier', coordinates=[[5.894436974354539, 51.991877507180675], [5.894429617405334, 51.991876221207356], [5.894393754011804, 51.99186950962195], [5.894378463265797, 51.991866654326714], [5.894378222251264, 51.99186658525656], [5.894377975524996, 51.991866526098256], [5.894377723107441, 51.991866478649314], [5.8943774649883185, 51.99186644201097], [5.894377202633767, 51.99186641707575], [5.894376938945331, 51.99186640293229], [5.8943766739434, 51.99186640137817], [5.8943764090635, 51.99186641060957], [5.89437614721739, 51.99186643061386], [5.894375886969502, 51.991866463195], [5.8943756312113225, 51.99186650654278], [5.894375381398706, 51.991866560650934], [5.894375138987564, 51.99186662551316], [5.894374903977898, 51.99186670112958], [5.894374679261083, 51.99186678568992], [5.894374463401657, 51.9918668809981], [5.894374259290994, 51.99186698524402], [5.894374068374837, 51.99186709752252], [5.894373889207444, 51.99186721873871], [5.894373726125989, 51.99186734617747], [5.894373577694954, 51.99186748164254], [5.894373443883829, 51.99186762243759], [5.89437332760428, 51.99186776855011], [5.894373228856464, 51.99186791998], [5.894373146164027, 51.99186807493605], [5.894373082438792, 51.991868233405704], [5.894373037670643, 51.9918683944901], [5.894364404111697, 51.99188656390106], [5.894427580929397, 51.991898003360305], [5.894436974354539, 51.991877507180675]]),
    ParkingSpot(spot_id=105, parking_type='bijzonder', street='Coehoornstraat', traffic_sign='E6a', neighborhood='Stationsplein', neighborhood_code='BU02020104', district=None, district_code='WK020201', area='26001', coordinates=[[5.901505793265879, 51.9828482239919], [5.901529780627893, 51.98285299710266], [5.901535071524492, 51.982843693492526], [5.901551042852021, 51.982814357574696], [5.901540361654022, 51.98279688596718], [5.901539118911973, 51.98279133659611], [5.901505793265879, 51.9828482239919]]),
    ParkingSpot(spot_id=300, parking_type='bijzonder', street='Van Lawick van Pabststraat', traffic_sign='E6a', neighborhood='Burgemeesterswijk', neighborhood_code='BU02021261', district=None, district_code='WK020212', area='Sector C Burgermeesterkwartier', coordinates=[[5.893895836335086, 51.98977695411586], [5.893853776972709, 51.989739324876886], [5.893809155142739, 51.989758237963656], [5.893850887776105, 51.98979607266101], [5.893895836335086, 51.98977695411586]]),
    ParkingSpot(spot_id=309, parking_type='bijzonder', street='Nieuwe Plein', traffic_sign='E6a', neighborhood='Rijnstraat', neighborhood_code='BU02020102', district=None, district_code='WK020201', area='26001', coordinates=[[5.902766117364105, 51.981544037872425], [5.902748904199828, 51.981507540574846], [5.9026998824497365, 51.98151646691924], [5.902717322352908, 51.98155281132695], [5.902766117364105, 51.981544037872425]]),
    ParkingSpot(spot_id=756, parking_type='bijzonder', street='Van Lawick van Pabststraat', traffic_sign='E6a', neighborhood='Burgemeesterswijk', neighborhood_code='BU02021261', district=None, district_code='WK020212', area='Sector C Burgermeesterkwartier', coordinates=[[5.893853776972709, 51.989739324876886], [5.893895836335086, 51.98977695411586], [5.893941581776814, 51.9897574977546], [5.89393613554183, 51.989752625260806], [5.893915669892394, 51.98973431606262], [5.893899572400147, 51.98971991325834], [5.893853776972709, 51.989739324876886]]),
    ParkingSpot(spot_id=807, parking_type='bijzonder', street='Oude Kraan', traffic_sign='E6a', neighborhood='Utrechtsestraat', neighborhood_code='BU02020170', district=None, district_code='WK020201', area='26001', coordinates=[[5.898927588531049, 51.98239198624151], [5.898939750346514, 51.98240702389143], [5.899049077518848, 51.98237326800506], [5.899044617100125, 51.982374006479475], [5.899039898973593, 51.9823738472393], [5.899006696963277, 51.98236759197505], [5.898927588531049, 51.98239198624151]]),
    ParkingSpot(spot_id=1170, parking_type='bijzonder', street='Stationsplein', traffic_sign='E6a', neighborhood='Utrechtsestraat', neighborhood_code='BU02020170', district=None, district_code='WK020201', area='26001', coordinates=[[5.899032949763431, 51.98389040528316], [5.899068282607165, 51.98388658790644], [5.899061386392674, 51.983862309806625], [5.899041777730751, 51.98379328456424], [5.899041579524576, 51.98379377439305], [5.899041331734853, 51.983794256348], [5.8990410372420445, 51.98379472772017], [5.899040697481224, 51.9837951867056], [5.8990403124420805, 51.98379563240539], [5.899039883559848, 51.98379606301581], [5.899039412269492, 51.98379647673285], [5.899016052760941, 51.98381726955005], [5.899015520263274, 51.98381780487598], [5.8990150535252415, 51.98381836238682], [5.899014651060418, 51.98381893939235], [5.899014318650261, 51.98381953227216], [5.899014054808342, 51.983820138336206], [5.899013863870841, 51.98382075486911], [5.899013745796666, 51.98382137827569], [5.899013700555044, 51.98382200585965], [5.899013728094739, 51.98382263312695], [5.899013828384916, 51.983823257381246], [5.899014002850458, 51.98382387591983], [5.899026040702751, 51.98386614791039], [5.899032949763431, 51.98389040528316]]),
    ParkingSpot(spot_id=1305, parking_type='bijzonder', street='Stationsplein', traffic_sign='E6a', neighborhood='Utrechtsestraat', neighborhood_code='BU02020170', district=None, district_code='WK020201', area='26001', coordinates=[[5.899068282607165, 51.98388658790644], [5.899032949763431, 51.98389040528316], [5.899034445031185, 51.98389565696483], [5.8990392769450875, 51.98391214665021], [5.899051379487812, 51.98395345031051], [5.899051573072502, 51.983954085843735], [5.899051834958903, 51.983954711192574], [5.899052166582208, 51.983955324553165], [5.899052563544673, 51.9839559232481], [5.899053025805316, 51.98395650368227], [5.899053550422051, 51.98395706317191], [5.899054135918759, 51.98395759992577], [5.899054779343146, 51.983958110361414], [5.899055477763436, 51.98395859269381], [5.899056226802461, 51.98395904604327], [5.899057023497665, 51.98395946592851], [5.8990900433925155, 51.98397389238287], [5.899090653899557, 51.983974197128155], [5.899091241301253, 51.9839745181531], [5.89909180415221, 51.98397485636268], [5.899092340976195, 51.98397520996572], [5.8990928503074604, 51.98397557806977], [5.89909333214589, 51.98397596067479], [5.899093783569973, 51.98397635689464], [5.899074428045405, 51.98390821950561], [5.899068282607165, 51.98388658790644]]),
    ParkingSpot(spot_id=1367, parking_type='bijzonder', street='Rijnkade', traffic_sign='E6a', neighborhood='Stationsplein', neighborhood_code='BU02020104', district=None, district_code='WK020201', area='26001', coordinates=[[5.900772747190894, 51.97996971843089], [5.900702768645455, 51.97992405026028], [5.900674716514379, 51.97994105285022], [5.900744864487038, 51.979986009318495], [5.900772747190894, 51.97996971843089]]),
    ParkingSpot(spot_id=2219, parking_type='bijzonder', street='Nieuwe Plein', traffic_sign='E6a', neighborhood='Rijnstraat', neighborhood_code='BU02020102', district=None, district_code='WK020201', area='26001', coordinates=[[5.902915226028621, 51.98141016242192], [5.902932913034013, 51.98144659110485], [5.902979918298003, 51.981437889224594], [5.9029619669067905, 51.98140151024486], [5.902915226028621, 51.98141016242192]]),
  ])
# ---
